nixos/openvswitch : upgrade 2.5.4 -> 2.9.2 and vlans

netixx opened the pull request:
Originnaly, the need was to fix #34336 and add features in #34851.

With remarks from @volth, openvswitch was upgraded to the latest
stable version (currenty 2.8.1).

Systemd dependencies were refactored according to analysis in #34586.

Vlans and internal interfaces are implemented following the
 `networking.interfaces` template format i.e. each interface is
described using an attributeSet (vlan and type at the moment).
If vlan is present, then interface is added to the vswitch with
given tag (access mode). Type internal enables `vswitchd` to create
interfaces (see openvswitch docs).

I am not sure about this implementation and this PR may need some tuning. Also, this change breaks compatibility with previous `vswitch.interfaces` syntax.

EDIT: More details on implementation:
I only included `type` and `vlan` options but there are a lot more (see `set` command in ovs-vsctl) and I hesitated to put a generic `key=value` options that would be translated in `set port/interface <key>=<value>`, but I think for usability for the user the minimal options are better. Maybe we could have a list of commands as an interface parameter, and we would replace a token by the name of the current interface in the actual commands (but I don't know how to do that).

About `type` argument, I originally thought (reading the docs) that it was required to apply this tag to virtual interfaces attached to the bridge, but actually, it leads to an error if the interface already exists (say created by `networking.interface` with option `virtual=true`). In the end, the 'type=internal` enables `vswitchd` to create the given interface. I think we could use this in a container scenario for example (by changing the network namespace of this interface), which is why I left it.

###### Motivation for this change
- Previous implementation would fail to start correctly.
- Version 2.5.2 would create a flap during startup which needed an ugly `sleep 1` to work
- Vlan feature is now implemented. It enables interfaces to be added in tagged/access mode

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

netixx committed:
nixos/openvswitch : upgrade 2.5.4 -> 2.9.2 and vlans

Originnaly, the need was to fix #34336 and add features in #34851.

With remarks from @volth, openvswitch was upgraded to the latest
stable version (currenty 2.9.2). This remove ovs-monitor-ipsec
commands. LTS version is still available using
`config.virtualisation.vswitch.package = pkgs.openvswitch-lts`

Systemd dependencies for scripted mode
were refactored according to analysis in #34586.

networking.vswitches can now be used with systemd-networkd,
although they are not supported by the daemon, a nixos receipe
creates the switch and attached required interfaces (just like
the scripted version).

Vlans and internal interfaces are implemented following the
 `networking.interfaces` template format i.e. each interface is
described using an attributeSet (vlan and type at the moment).
If vlan is present, then interface is added to the vswitch with
given tag (access mode). Type internal enabled vswitch to create
interfaces (see openvswitch docs).

Added configuration for configuring supported openFlow version on
the vswitch

netixx commented:
I saw while rebasing that the upgrade to 2.5.4 was performed by @jokogr indeed!

I couldn't test for the setup bug with this version, but 2.8.1 brings interesting stuff (in the IPSec department for example).

jokogr commented on pkgs/os-specific/linux/openvswitch/default.nix:
```diff
@@ -36,26 +37,27 @@ in stdenv.mkDerivation rec {
-    substituteInPlace $out/share/openvswitch/scripts/ovs-monitor-ipsec \
-      --replace "self.cert_dir" "root_prefix + self.cert_dir"
-  '';
+  # ovs-monitor-ipsec removed in ovs: 2b02d770c4cb381ec32cd4b7b1e991c42b448884
```
No need to leave this as a comment since the command was removed

netixx replied:
Ok. I wanted to make sure I didn't forget some other script in this step. Should I leave the comment about removal of the script in upstream ?

jokogr replied:
IMHO no, since it is now removed and users should read the OVS documentation before using a new version.

netixx replied:
done !

jokogr commented:
To be honest, I was using `extraOvsctlCmds` in order to tag ports so far, as well as a minor modification in network-interfaces-scripted.nix to ensure the tagged interfaces were starting after the actual ones.

This one seems much cleaner, let me try in for a couple of days (I am heavily modifying VLANs etc.) before totally approving it.

netixx commented:
I thought about `extraOvsctlCmds` too but I didn't want to repeat interface names in multiple places...

Thank you for looking at this ! I have to play with the `slaves` variables a bit to make it work (e.g. for `dhcpcd`, if the attached interface is a virtual or internal then we should be able to do dhcp on it).

volth commented:
about ipsec: https://github.com/NixOS/nixpkgs/pull/24498#issuecomment-295020283
that code never worked well

netixx commented:
I did some more testing in a more complex scenario and found a few snags in the current implementation, I originally wanted to attach tap (using `virtual=true` and `virtualType=tap`) devices to the bridges (in vlans) and assign IP to them, but this does not work [see here](http://docs.openvswitch.org/en/latest/faq/issues/). Instead, we should let openvswitch create the interfaces (which is what `type=internal` does) and assign IP addresses on those interfaces.

I worked on a new implementation:
- I needed to add `before` dependency to `network-addresses-${i}.service` for all internal interfaces because otherwise, ip address assignment failed (`Cannot find device` in the start script for addresses).
- I removed the IPsec related config options (which referenced the script that was removed in 2.8)
- I added an schema upgrade script in ovsdb (I got warnings when testing after upgrade to 2.8)
- the binding to the subsystem devices was killing the service upon startup (during the pre-start), so I removed it for now. I don't think systemd supports binding to the subsystem when your are deleting/creating said device. Perhaps we should split the service (with creation of the bridge in a separate unit)

I have a couple of questions still:
- Openvswitch 2.9 is out, should I push the upgrade as well ?
- I saw `db = pkgs.stdenv.mkDerivation {` in openvswitch.nix, but I don't think it's used for the database (I am not sure though)
- What about `subsystemDevice` binding ?

I will push the upgrade soon (after some more testing on my side).

netixx commented:
Since we are on the subject of openvswitch, do you know of limitations regarding usage of `systemd-networkd` with `networking.vswitches` (I know it is forbidden by nixos, but I don't know why) ?

(I am planning on supporting both `networking.useNetworkd` and `networking.vswitches` at the same time because I need it for my setup)

volth commented:
> Since we are on the subject of openvswitch, do you know of limitations regarding usage of systemd-networkd with networking.vswitches (I know it is forbidden by nixos, but I don't know why) 

AFAIK, ```systemd-network``` does not support ovs: https://github.com/systemd/systemd/issues/2613
But I do not know how it could be forbidden by nixos

netixx commented:
Thank you for this info. I was just thinking of configuring the bridge with a nixos generated unit, and then let systemd-networkd do the address and other parts. The way I see it, I just have to insert this config unit before the systemd-network process (and maybe prevent the physical interfaces from matching networkd).

netixx commented:
I pushed an implementation in the `network-interfaces-systemd.nix` which means that `networking.vswitches` can now be used with `networking.useNetworkd=true` :)

I found that using systemd is much, much, much more stable than scripted interfaces. Internal interfaces (created by openvswitch) can be referred to and configured in the normal `networking.interfaces`. The only thing that could need fixing (improved, really), is to add a warning if the user tries to configure a IP address and a physical (i.e. **not** internal) interface : this actually works, but the address will be unusable.

Also, I fixed an mistake in the stop process of the switch (that was not cleaning the switch correctly)

jokogr commented:
@netixx this is impressive, thank you for your work!  I will try the module with systemd-networkd this week.

So far setting VLANs with your module (and not `extraOvsctlCmds`) is working as expected for my workloads.

Version 2.8.2 seems also stable to me, but I have a small request to make: Is it possible to keep the LTS version as well, and allow users to select it via a `package` attribute (unsure where to put it, though)?

netixx commented:
Including both version seems the most flexible choice to me too !

I am not sure how we should implement (I am not very familiar with package building...), do we include a `lts` flag on the package ? What should the default value be (false ?) ?

netixx commented:
Did you have time to look at the new implementation ?

fpletz commented on nixos/modules/tasks/network-interfaces.nix:
```diff
@@ -487,9 +513,8 @@ in
 
           interfaces = mkOption {
             example = [ "eth0" "eth1" ];
```
Are you sure the example still compatible with the new type?

netixx replied:
Good catch! I forgot to update the example.

jokogr reviewed (approved):
Many thanks for including the LTS version!

netixx commented:
I resolved the conflict (by re-adding the IPSec behaviour only for lts version).

I also added the option to select the supported openFlow version for the switch.

netixx commented:
Are there still things I should do to move this along ?

jokogr commented:
@netixx could you please rebase to master and upgrade both LTS (v2.5.5) and latest (v2.9.2) Open vSwitch packages? I promise to test both.

Your contributions are very useful, I want to see them upstream, too.

jokogr commented:
@netixx I have been using your changes for more than 2 months with the following config:

```
  virtualisation.vswitch = {
    enable = true;
    resetOnStart = true;
    package = pkgs.openvswitch-lts;
  };

  networking.vswitches.vs0 = {
    interfaces = {
      "eno1" = {};
      "mgmt0" = {
        type = "internal";
      };
    };
  };

  networking.interfaces.mgmt0 = {
    ipv4.addresses = [{
      address = "192.168.1.10";
      prefixLength = 24;
    }];
  };

  networking.defaultGateway = "10.101.9.1";

  networking.nameservers = [ "1.1.1.1" ];
```

The problem with that setup is that **after a while** (so in the beginning / after a reboot everything is fine) the machine becomes unreachable via SSH, although I can ping it. Do you have any idea what it could be wrong?

netixx commented:
If you can still ping, this means that the IP configuration is still working. I've had problems with dhcpd: when I changed the bridge configuration (eg. adding an interface), running `nixes-rebuild switch` would restart the bridge (erase & recreate) and dhcpd would fail to keep the binding to the interfaces (because they are removed and recreated).

I also found that using `networking.useNetworkd = true` worked best. I couldn't make the scripted interfaces work properly (I think the dependencies are messed up).

Do you have a specific configuration in ssh (such as bind address for example) ?
Do you have access to the machine via a console (to issue debug commands) ?

jokogr commented:
@netixx ssh setup was simple, no interface binding took place and unfortunately this is a headless setup with no other way to access it after the lockout.

I might set up a VM to check its behaviour, but for the time being I have followed your suggestion to use networkd and so far, so good.

In my previous setup I had a static IP address, but the NixOS module does create by default an internal port with the bridge's name, so dhcpd might have causing the problem.

Anything else I could do to get this merged?

netixx commented:
> Anything else I could do to get this merged?

We should attract the attention of the maintainers, but I don't know how...

jokogr commented:
@netixx hey, could you split your PR into two? One for the LTS version and package updates and another for the service improvement.

I could do it if you want me to. I would like to eventually use the latest OVS with your declarative configuration.
